### DISCUSSION TITLE
IfI start a new project with PyO3, will maturin restrict my choices later?

### DISCUSSION THREAD

**Discussion by haberman** [PyO3/pyo3#1915] [Questions] [answered]
I'm brand new to pyo3. In the README I see:

> PyO3 can be used to generate a native Python module. The easiest way to try this out for the first time is to use maturin. maturin is a tool for building and publishing Rust-based Python packages with minimal configuration.

The language here suggests that maturin has some limitations, but it doesn't help me gauge whether I will run into those limitations or not. I want to avoid investing into a tool that I will soon outgrow.

It would be very helpful to see a compare/contrast between maturin, setuptools-rust, and manual builds. What use cases require the more complicated tools?

**Comment by mejrs** [comment, marked as answer]
There's some more documentation in the guide: https://pyo3.rs/v0.14.5/building_and_distribution.html#packaging-tools

There's not really a cost to switching between them or supporting both, so don't worry about "making the wrong choice early on".

> What use cases require the more complicated tools?

See https://github.com/InstantDomain/instant-segment/pull/24 for an example of that.
